### DISCUSSION TITLE
Why does the CSRF class have an additional `secret` option?

### DISCUSSION THREAD

**Discussion by KingSora** [sergiodxa/remix-utils#310] [Q&A] [answered]
Good day :)

I'm a little bit confused with the current `v7` `CSRF` class and its optional `secret` option.
Why is it there? If the passed cookie already has an specified `secrets` array, specifying the option would result in double signing.
Migrating from `v6` with this option set would break already set cookies because reading the old cookie would result in an error since it expects a double signed cookie.

Could you clarify for me why the `secret` option is present and what its intentions are?

**Comment by sergiodxa** [comment, marked as answer]
In case you want to sign the value without signing the cookie, remember that the value is also sent to the browser, not just stored in a cookie.

If you don't set it it will not be used at all so your previous cookies will keep working.

**Reply by KingSora** [reply]
@sergiodxa thanks for your response :)

Correct me if I'm wrong, but "signing the cookie" also just signs the cookie value (https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/cookies.ts#L110-L128) so there should be no difference between "signing the cookie" and "signing the value"

Which leads me again to the question why offering the separate signing functionality if `remix` itself and its cookie abstraction already have an implementation for it. (even an slightly better one because one can specify an array of secrets)

**Reply by sergiodxa** [reply]
Signing a cookie doesn't sign the value, because if you do this:
```ts
let value = getValue()
return json({ value }, {
  headers: { "set-cookie": await cookie.serialize(value) }
})
```

The `value` returned in the response body is not signed, only the cookie is.

If you don't sign your cookie and don't want to start doing it you can sign the value of the CSRF token itself.

Anyway, just an option surfaced since the crypto library offered it.
